src/validation.rs

```diff
@@ -37,11 +37,40 @@ impl<Domain: Display> Display for ValidationErr<Domain> {
     }
 }
 
-pub trait Validate<Domain: Display> {
+pub trait Validate<RuntimeDomain: Display, ContractDomain: Display> {
```

Comment:
Looks like Validate could be decomposed into two traits:
```
pub trait ValidateRuntimeCall {
fn validate(..
}

pub trait ValidateContractCall {
fn validate(..
}
```
This way, the separation of the returned error type could make more sense and also there could be no redundant validate function.